File: sdk-typescript/src/beliefs/divergence.ts
/**
 * Belief divergence detection — pure-function module.
 *
 * Ported from sdk-python/synapse/beliefs/divergence.py. No I/O, fully
 * testable. When multiple agents assert different values for the same
 * belief key, the divergence detector flags it.
 */

export type BeliefSource = "observed" | "inferred" | "assumed" | string;

export interface AgentBelief {
  agent_id: string;
  key: string;
  value: unknown;
  confidence: number;
  /** "observed" | "inferred" | "assumed" */
  source: BeliefSource;
}

export interface BeliefDivergence {
  key: string;
  agents: AgentBelief[];
  severity: number;
  /** Distinct values held across the agents (insertion order). */
  distinct_values: unknown[];
}

/** Numeric weight derived from `source` × `confidence`. */
export function evidentialWeight(b: AgentBelief): number {
  const rankMap: Record<string, number> = {
    observed: 1.0,
    inferred: 0.7,
    assumed: 0.4,
  };
  const rank = rankMap[b.source] ?? 0.5;
  const w = b.confidence * rank;
  if (Number.isNaN(w)) return 0.0;
  return Math.min(1.0, Math.max(0.0, w));
}

/** Float-fuzz comparison mirroring Python's _values_equal. */
export function valuesEqual(a: unknown, b: unknown): boolean {
  if (
    typeof a === "number" &&
    typeof b === "number" &&
    !Number.isInteger(a) &&
    !Number.isInteger(b)
  ) {
    return Math.abs(a - b) < 1e-9;
  }
  // Mirror Python equality for floats vs ints — if either is float-typed by
  // having a fractional part, use fuzz; otherwise structural equality.
  if (typeof a === "number" && typeof b === "number") {
    return Math.abs(a - b) < 1e-9;
  }
  // Structural deep-equal for arrays/objects, primitive equality for scalars.
  if (a === b) return true;
  if (a === null || b === null) return false;
  if (typeof a !== typeof b) return false;
  if (typeof a === "object") {
    try {
      return JSON.stringify(a) === JSON.stringify(b);
    } catch {
      return false;
    }
  }
  return false;
}

/** Compute distinct values within a group, preserving insertion order. */
function distinctIn(group: AgentBelief[]): unknown[] {
  const distinct: unknown[] = [];
  for (const b of group) {
    if (!distinct.some((d) => valuesEqual(b.value, d))) {
      distinct.push(b.value);
    }
  }
  return distinct;
}

/**
 * Group beliefs by key; emit a BeliefDivergence whenever ≥2 agents disagree
 * on the same key. Returned list is sorted severity-descending.
 */
export function detectDivergences(
  beliefs: Iterable<AgentBelief>,
): BeliefDivergence[] {
  const byKey = new Map<string, AgentBelief[]>();
  for (const b of beliefs) {
    let arr = byKey.get(b.key);
    if (!arr) {
      arr = [];
      byKey.set(b.key, arr);
    }
    arr.push(b);
  }

  const out: BeliefDivergence[] = [];
  for (const [key, group] of byKey) {
    if (group.length < 2) continue;
    const distinct = distinctIn(group);
    if (distinct.length < 2) continue;
    const avgWeight =
      group.reduce((s, b) => s + evidentialWeight(b), 0) / group.length;
    const scale = Math.min(1.0, distinct.length / 3.0);
    const severity = Math.min(1.0, avgWeight * (0.5 + 0.5 * scale));
    out.push({
      key,
      agents: [...group],
      severity,
      distinct_values: distinct,
    });
  }

  out.sort((a, b) => b.severity - a.severity);
  return out;
}

/** Build AgentBelief[] from generic DB-row dicts. */
export function beliefsFromDbRows(
  rows: Iterable<Record<string, unknown>>,
): AgentBelief[] {
  const out: AgentBelief[] = [];
  for (const r of rows) {
    out.push({
      agent_id: String(r["agent_id"]),
      key: String(r["key"]),
      value: r["value"],
      confidence: Number(r["confidence"]),
      source: String(r["source"]),
    });
  }
  return out;
}

File: sdk-python/synapse/beliefs/divergence.py
"""Belief divergence detection — pure-function module.

Mirrors ``runtime.coordinator.belief_divergence`` so the SDK doesn't
need the runtime/ directory on its import path. The SDK is the
source of truth; the runtime can re-export from here in a future cleanup.

When multiple agents assert different values for the same belief key,
the divergence detector flags it.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Iterable


@dataclass(frozen=True)
class AgentBelief:
    agent_id: str
    key: str
    value: Any
    confidence: float
    source: str   # "observed" | "inferred" | "assumed"

    @property
    def evidential_weight(self) -> float:
        rank = {"observed": 1.0, "inferred": 0.7, "assumed": 0.4}.get(self.source, 0.5)
        return min(1.0, max(0.0, self.confidence * rank))


@dataclass(frozen=True)
class BeliefDivergence:
    key: str
    agents: tuple[AgentBelief, ...]
    severity: float

    @property
    def distinct_values(self) -> tuple[Any, ...]:
        seen: list[Any] = []
        for b in self.agents:
            if b.value not in seen:
                seen.append(b.value)
        return tuple(seen)


def _values_equal(a: Any, b: Any) -> bool:
    if isinstance(a, float) and isinstance(b, float):
        return abs(a - b) < 1e-9
    return a == b


def detect_divergences(beliefs: Iterable[AgentBelief]) -> list[BeliefDivergence]:
    by_key: dict[str, list[AgentBelief]] = {}
    for b in beliefs:
        by_key.setdefault(b.key, []).append(b)

    out: list[BeliefDivergence] = []
    for key, group in by_key.items():
        if len(group) < 2:
            continue
        distinct: list[Any] = []
        for b in group:
            if not any(_values_equal(b.value, d) for d in distinct):
                distinct.append(b.value)
        if len(distinct) < 2:
            continue
        avg_weight = sum(b.evidential_weight for b in group) / len(group)
        scale = min(1.0, len(distinct) / 3.0)
        severity = min(1.0, avg_weight * (0.5 + 0.5 * scale))
        out.append(BeliefDivergence(key=key, agents=tuple(group), severity=severity))

    out.sort(key=lambda d: d.severity, reverse=True)
    return out


def beliefs_from_db_rows(rows: Iterable[dict[str, Any]]) -> list[AgentBelief]:
    out: list[AgentBelief] = []
    for r in rows:
        out.append(AgentBelief(
            agent_id=r["agent_id"],
            key=r["key"],
            value=r["value"],
            confidence=float(r["confidence"]),
            source=r["source"],
        ))
    return out
